fedcloud_monitoring_tools/vo_mon.py

```diff
@@ -0,0 +1,157 @@
+
+    def create_vm_tosca_template(self):
+        inf_desc = """
+            tosca_definitions_version: tosca_simple_yaml_1_0
+            imports:
+            - grycap_custom_types: https://raw.githubusercontent.com/grycap/tosca/main/custom_types.yaml
+            topology_template:
+              node_templates:
+                simple_node:
+                  type: tosca.nodes.indigo.Compute
+                  capabilities:
+                    endpoint:
+                      properties:
+                        network_name: PUBLIC
+                    host:
+                      properties:
+                        num_cpus: 2
+                        mem_size: 4 GB
+                    os:
+                      properties:
+                        image: appdb://SITE/egi.ubuntu.24.04?VO
+              outputs:
+                node_ip:
+                  value: { get_attribute: [ simple_node, public_address, 0 ] }
+                node_creds:
+                  value: { get_attribute: [ simple_node, endpoint, credential, 0 ] }
+        """
+        return inf_desc.replace("SITE", self.site).replace("VO", self.vo)
+
+    def launch_test_vm(self, ssh_command):
+        # deploy VM
+        tosca_template = self.create_vm_tosca_template()
+        self.create_auth_file(AUTH_FILE)
+        auth = IMClient.read_auth_data(AUTH_FILE)
+        imclient = IMClient.init_client(IM_REST_API, auth)
+        click.echo("[+] Creating test VM...")
+        success, inf_id = imclient.create(tosca_template, desc_type="yaml")
+        if not success:
+            raise VOTestException(inf_id)
+        click.echo(f"[+] Test VM successfully created with ID {inf_id}")
+        # wait for VM to be ready
+        state = "pending"
+        attempts = 1
+        while state != "configured" and attempts <= MAX_NUM_CHECKS:
+            click.echo(
+                f"[+] Waiting for test VM to be ready. Current state is: {state}. Attempt: {attempts}/{MAX_NUM_CHECKS}"
+            )
+            time.sleep(10)
+            success, state = imclient.getvminfo(inf_id, 0, prop="state")
+            attempts += 1
+        # has the VM been configured?
+        if state != "configured":
+            click.secho(
+                f"Test VM could not be configured after {attempts} attempts, reported state is: {state}",
+                fg="red",
+                bold=True,
+            )
+            success, state = imclient.getvminfo(inf_id, 0, prop="contmsg")
+            click.echo(f"Further information about the VM: {state}")
+            self.destroy_test_vm(inf_id)
+            return False
+
+        click.echo(
+            f"[+] Test VM is now {state}. Waiting for 1 additional minute before testing the VM."
+        )
+        time.sleep(60)
+        # run SSH command inside the VM
+        success, outputs = imclient.get_infra_property(inf_id, "outputs")
+        if not success:
+            raise VOTestException(f"{success} {outputs}")
```

Comment:
should the VM be destroyed in this case?

Reply:
actually, on second thoughts, this `if` statement can be deleted; it doesn't provide much valuable info after all.

Reply:
resolving comment to merge, thanks!